lib/THC/THCTensorCopy.cu

```diff
@@ -7,5 +7,226 @@ inline int curGPU() {
   return curDev;
 }
 
+// Copy operator for the pointwise apply kernel
+template <typename TypeDst, typename TypeSrc>
+struct CopyOp {
+  __device__ __forceinline__ void operator()(TypeDst* dst, TypeSrc* src) {
+#if __CUDA_ARCH__ >= 350
+    *dst = ScalarConvert<TypeSrc, TypeDst>::to(__ldg(src));
+#else
+    *dst = ScalarConvert<TypeSrc, TypeDst>::to(*src);
+#endif
+  }
+};
+
+// Copy for the same type to the same type
+template <typename TensorTypeDst, typename TensorTypeSrc>
```

Comment:
Most everything else in this diff is pretty mechanical; I think it would work if it compiled (excepting the half math implementations, which I may have messed up).

This function contains the one piece of real functionality change, since I now handle GPU -> GPU copies between different types exclusively on the GPU using this copy function.

This copy function is based off of the old GPU -> GPU copy function, but adds a case where the type of tensor differs between src and dst.

If there is one function worth looking at in detail and comparing against the old copy function, it is this one.